src/core_test/mpas_test_openacc.F

```diff
@@ -0,0 +1,362 @@
+
+        nullify(array_cpu)
+        call mpas_pool_get_array(openaccTest_pool, 'edge_cpu', array_cpu)
+
+        nullify(array_gpu)
+        call mpas_pool_get_array(openaccTest_pool, 'edge_gpu', array_gpu)
+
+        call rep_arrs_work_cpu(nCells,nEdges,nCellsSolve,nEdgesSolve, &
+                               areaCell,indexToCellID,nEdgesOnCell,cellsOnEdge, &
+                               array_cpu)
+
+        call rep_arrs_work_gpu(nCells,nEdges,nCellsSolve,nEdgesSolve, &
+                               areaCell,indexToCellID,nEdgesOnCell,cellsOnEdge, &
+                               array_gpu)
+
+        call diff_arrs(nEdges, array_cpu, array_gpu, diff)
+
+        if (diff > 0.0_RKIND) then
+            ierr = ierr + 1
+        end if
+
+        call mpas_dmpar_max_int(domain % dminfo, ierr, ierr_global)
+        ierr = ierr_global
+
+    end function openacc_test_rep_arrs
+
+
+    !-----------------------------------------------------------------------
+    !  routine rep_arrs_work_cpu
+    !
+    !> \brief CPU work routine for OpenACC representative arrays test
+    !> \author G. Dylan Dickerson
+    !> \date   29 May 2024
+    !> \details
+    !> Performs some array work on the CPU, based on patterns in the
+    !> MPAS-A dycore.
+    !
+    !-----------------------------------------------------------------------
+    subroutine rep_arrs_work_cpu(nCells, nEdges, nCellsSolve, nEdgesSolve, &
+                                 areaCell, indexToCellID, nEdgesOnCell, cellsOnEdge, &
+                                 edge_arr_cpu)
+
+        use mpas_kind_types, only : RKIND
+
+        implicit none
+
+        ! arguments
+        integer, intent(in) :: nCells, nEdges, nCellsSolve, nEdgesSolve
+        real (kind=RKIND), dimension(:), intent(in) :: areaCell
+        integer, dimension(:), intent(in) :: indexToCellID
+        integer, dimension(:), intent(in) :: nEdgesOnCell
+        integer, dimension(:,:), intent(in) :: cellsOnEdge
+        real (kind=RKIND), dimension(:), intent(inout) :: edge_arr_cpu
+
+        ! locals
+        integer :: iCell, iEdge, cell1, cell2
+        real (kind=RKIND), dimension(nCells) :: invArea, help_arr
+
+        ! Compute any helpers and initialize arrs
+        do iCell=1,nCells
+           invArea(iCell) = 1.0_RKIND / areaCell(iCell)
+           help_arr(iCell) = 0.0_RKIND
+        end do
+        do iEdge=1,nEdges
+           edge_arr_cpu(iEdge) = 0.0_RKIND
+        end do
+
+        ! Compute helper values (for all owned cells)
+        do iCell=1,nCellsSolve
+           help_arr(iCell) = (nEdgesOnCell(iCell)+indexToCellID(iCell)) * invArea(iCell)
+        end do
+
+        ! Compute final value (for all owned edges)
+        do iEdge=1,nEdgesSolve
+           cell1 = cellsOnEdge(1,iEdge)
+           cell2 = cellsOnEdge(2,iEdge)
+
+           edge_arr_cpu(iEdge) = 0.5_RKIND * (help_arr(cell1) + help_arr(cell2))
```

Comment:
It's not necessarily a problem since all cells (owned and halo) are initialized in the `help_arr` array, but note that there may be owned edges that form the boundary between an owned cell and a halo cell.

Reply:
That crossed my mind too, so I was careful to initialize everything possible. An extension of these tests could include a GPU-direct halo exchange and some operation to "advect" those halo values.